src/lua/compat.lua

```diff
@@ -0,0 +1,294 @@
+    result.default = options[key].default
+
+    if options[key].obsolete then
+        result.obsolete = true
+    end
+
+    if not cfg[key].selected then
+        result.current = 'default'
+    elseif cfg[key].current == NEW then
+        result.current = 'new'
+    else
+        result.current = 'old'
+    end
+
+    return result
+end
+
+-- Returns info on all unobsolete options from `options`.
+local
+function serialize_compat()
+    local result = { }
+    local cnt = 1
+    for _, name in pairs(options_order) do
+        if cfg[name].selected and cfg[name].current == NEW and not options[name].obsolete then
+            result[cnt] = {[name] = 'new'}
+            cnt = cnt + 1
+        end
+    end
+    for _, name in pairs(options_order) do
+        if cfg[name].selected and cfg[name].current == OLD and not options[name].obsolete then
+            result[cnt] = {[name] = 'old'}
+            cnt = cnt + 1
+        end
+    end
+    for _, name in pairs(options_order) do
+        if not cfg[name].selected and not options[name].obsolete then
+            result[cnt] = {[name] = ('default (%s)'):format(cfg[name].current and 'new' or 'old')}
+            cnt = cnt + 1
+        end
+    end
+    return result
+end
+
+-- Checks options correctness and sets `option.default` to boolean value if needed.
+local function verify_option(name, option)
+    if type(name) ~= 'string' then
+        error(('Option name must be a string (%s provided)'):format(type(name)))
+    end
+    for p, t in pairs(options_format) do
+        if not (p == 'obsolete' and option[p] == nil) and type(option[p]) ~= t then
+            error(('Invalid option table for %s, bad %s (%s is expected)'):format(name, p, t))
+        end
+    end
+    if option.default ~= 'new' and option.default ~= 'old' then
+        error(("Invalid option table for %s, bad default ('new'/'old' is expected)"):format(name))
+    end
+
+    if option.obsolete and option.default == 'old' then
+        error(('Obsolete option %s default is wrong.'):format(name))
+    end
+end
+
+-- Checks if operation is valid, sets value to an option and runs postaction.
+local function set_option(name, val)
+    local option = options[name]
+    if not option then
+        error(('Invalid option %s'):format(name))
+    end
+    local selected;
+    if val == 'new' then
+        val = NEW
+        selected = true
+    elseif val == 'old' then
+        val = OLD
+        selected = true
+    elseif val == 'default' then
+        val = option.default == 'new'
+        selected = false
+    else
+        error(('Invalid argument %s'):format(val))
+    end
+    if not option.obsolete then
+        postaction[name](val)
+    elseif selected and val == OLD then
+        error('Chosen option is no longer available')
+    end
+    cfg[name].current = val
+    cfg[name].selected = selected
+end
+
+local compat = setmetatable({
+    candidates = function()
+        local result = { }
+        for key, val in pairs(options) do
```

Comment:
There are too many global tables (5) for it to look nice, so I decided to leave it in lowercase.

Reply:
It is not a global, it is an upvalue (what is called `nonlocal` in Python, a captured variable). Moreover, it is mutable.

I heard the following about historical reasons of uppercasing constants. Initially it was applied to macros in the C language, because macros are dangerous. The only way to declare a constant in K&R C is to define a macro (`enum` was added in ANSI C), so constants are used to be in the uppercase. It becomes so usual that goes to ANSI C as well and spread to constants that are defined using `enum`. However there is the opposite opinion on this: macros should be in the uppercase, constants shouldn't. Not very popular opinion, but still. (I didn't check this history, but if so, it is a fun fact.)

Anyway, global/upvalue/local in Lua usually has nothing to do with lower/uppercasing. Mutability/constantness does.

I agree that the `options` term is quite common and may look like a local value. We can name it `registered_options` or `compat_options`, but just `options` is okay for me too.